### PR TITLE
Add a hideConstantImplementations dartdoc directive

### LINKED ISSUE #2657: Hide implementation of static constants
The [Icons](https://api.flutter.dev/flutter/material/Icons-class.html) class defines icons as `IconData` constants. This has lead a number of devs to use the implementation (e.g. `IconData(57399, fontFamily: 'MaterialIcons')`) rather than the interface `Icons.airplane`. This is a problem because all codepoints (e.g. 57399) change every time the font is updated. 

Is there an option or annotation to hide the implementation in docs?

### PR DESCRIPTION
**Pull request by jcollins-g**
Fixes #2657.  

Without trying to solve #3397 this PR does also move a few things around to make it more clear where the various directives are currently implemented as adding yet another one via `buildDocumentationAddition` was starting to look messy.

Here is a sample screenshot of what `Icons` would look like with the new directive:

Screenshot 2023-04-25 at 2 05 39 PM

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -412,6 +412,41 @@ markdown link isn't linked).
...
+place the directive in the class documentation where the constants are defined.
+For top level constants, place the directive in the library where the constants
+are defined.
+
+For example:
+
+```dart
+/// This is truly an amazing library.
+/// {@hideConstantImplementations}
+library my_library;
```
Comment (srawlins) [inline README.md:431]: nit: I think we're steering away from named libraries like this; we can use `library;` in Dart 2.19.

File: lib/src/model/directives/canonical_for.dart [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:dartdoc/src/model/documentation_comment.dart';
+import 'package:dartdoc/src/model/library.dart';
+
+final _canonicalRegExp = RegExp(r'{@canonicalFor\s([^}]+)}');
+
+/// Used by [Library] to implement the canonicalFor directive.
```
Comment (srawlins) [inline lib/src/model/directives/canonical_for.dart:10]: Can we wrap 'canonicalFor' in some delimiter like backticks?

File: lib/src/model/directives/categorization.dart [hunk 3]
```diff
@@ -9,11 +9,11 @@ final RegExp _categoryRegExp = RegExp(
     r'[ ]*{@(api|category|subCategory|image|samples) (.+?)}[ ]*\n?',
     multiLine: true);
 
-/// Mixin implementing dartdoc categorization for ModelElements.
-mixin Categorization implements ModelElement {
+/// Mixin parsing the `@category` directive for ModelElements.
+mixin Categorization on DocumentationComment implements Indexable {
```
Comment (srawlins) [inline lib/src/model/directives/categorization.dart:13]: Nice cleanup!

File: lib/src/model/directives/hide_constant_implementations.dart [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    return _hideConstantImplementations!;
+  }
+
+  /// Hides [hideConstantImplementations] from doc while leaving a note to
+  /// ourselves to change rendering for these constants.
+  /// Example:
+  ///
+  ///     {@hideConstantImplementations}
+  @override
+  String buildDocumentationAddition(String rawDocs) {
```
Comment (srawlins) [inline lib/src/model/directives/hide_constant_implementations.dart:31]: With this implementation, do we need to parse every doc comment _n_ times given _n_ directives? (Assuming we ask each element questions like `hideConstantImplementations` once each element)

File: lib/src/model/directives/hide_constant_implementations.dart [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    RegExp(r'{@hideConstantImplementations}');
+
+/// Implement parsing the hideConstantImplementations dartdoc directive
+/// for this [ModelElement].  Used by [Container].
+mixin HideConstantImplementations on DocumentationComment {
+  bool? _hideConstantImplementations;
+
+  /// [true] if the {@hideConstantImplementations} dartdoc directive is present
+  /// in the documentation for this class.
+  bool get hideConstantImplementations {
```
Comment (srawlins) [inline lib/src/model/directives/hide_constant_implementations.dart:18]: The name here makes it sound like a function, like calling `hideConstantImplementations` will do something or set something. What about `hasHideConstantImplementations`, like how the analyzer APIs are named?

https://dart.dev/guides/language/effective-dart/design#prefer-a-non-imperative-verb-phrase-for-a-boolean-property-or-variable

File: lib/src/model/directives/hide_constant_implementations.dart [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    if (_hideConstantImplementations == null) {
+      buildDocumentationAddition(documentationComment);
+    }
+    return _hideConstantImplementations!;
+  }
+
+  /// Hides [hideConstantImplementations] from doc while leaving a note to
```
Comment (srawlins) [inline lib/src/model/directives/hide_constant_implementations.dart:25]: Hmm, is this reference right? This function doesn't hide the bool getter, `hideConstantImplementations`. Maybe "Hides `{@hideConstantImplementations}` ..."?

File: lib/src/model/documentation_comment.dart [hunk 7]
```diff
@@ -97,14 +97,11 @@ mixin DocumentationComment on Documentable, Warnable, Locatable, SourceCode {
   /// result.
   String _processCommentWithoutTools(String documentationComment) {
     var docs = stripComments(documentationComment);
-    if (!docs.contains('{@')) {
-      _analyzeCodeBlocks(docs);
-      return docs;
+    if (docs.contains('{@')) {
```
Comment (srawlins) [inline lib/src/model/documentation_comment.dart:100]: This changes causes us to execute `_stripHtmlAndAddToIndex` for docs which do not contain `{@`. But we know such docs don't contain `{@inject-html`, so it seems there is no reason to do that work?

File: lib/src/model/directives/hide_constant_implementations.dart (hunk 4 above)
Comment (srawlins) [reply lib/src/model/directives/hide_constant_implementations.dart:31]: Let's benchmark this then to see that another parse of each doc comment doesn't slow down, e.g. flutter doc generation.

Comment (srawlins) [reply lib/src/model/directives/hide_constant_implementations.dart:31]: We can also add this to an issue I have regarding duplicate parsing of comments: https://github.com/dart-lang/dartdoc/issues/2970

Comment (jcollins-g) [reply lib/src/model/directives/hide_constant_implementations.dart:31]: flutter doc generation timing:

new:  2492.9 seconds
old: 2518.7 seconds

~1% difference, within margin of error (I don't think this code actually improves things, so this is definitely a margin-of-error situation).   As this was running I realized the performance impact isn't as bad as I first implied.   We are increasing the number of parse-throughs per documentation comment, but this only increases the amount of time for:

1) For Containers (Class, Mixin, Enum), only those that contain at least one constant will be impacted.
2) For every element type, only those that contain at least one documentation comment will have this impact (there's a shortcut taken via a regex).   This overlaps with Containers, of course.

This explains why the measurable impact is negligible here.   It's still a code smell, though, and we should fix this longer term.

Comment (srawlins) [reply lib/src/model/directives/hide_constant_implementations.dart:31]: Yeah I'm not surprised. Good news. Thanks for checking!

File: lib/src/model/documentation_comment.dart (hunk 7 above)
Comment (jcollins-g) [reply lib/src/model/documentation_comment.dart:100]: done

File: lib/src/model/getter_setter_combo.dart [hunk 8]
```diff
@@ -266,6 +267,10 @@ mixin GetterSetterCombo on ModelElement {
 
   bool get writeOnly => hasPublicSetter && !hasPublicGetter;
 
+  /// True if the @hideConstantImplementations directive is present
+  /// in the defining enclosing element.
+  bool get hideConstantImplementation;
```
[the remark this answers is not in the document]
Comment (jcollins-g) [reply lib/src/model/getter_setter_combo.dart]: done

File: lib/src/model/directives/hide_constant_implementations.dart (hunk 6 above)
Comment (jcollins-g) [reply lib/src/model/directives/hide_constant_implementations.dart]: done

**Review by srawlins — APPROVED** [review]

**Pull request merged**